File: MtxVec.Delphi/DewProbCalc.pas
unit DewProbCalc;

interface

uses
  Windows, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, DewProbWrapper, StdCtrls, ExtCtrls, TeeProcs, TeEngine,
  Chart, Series, ComCtrls, Buttons, ActnList, ProbEditor, Math387, XPMan,
  ToolWin, ActnMan, ActnCtrls, XPStyleActnCtrls, StdActns, ActnColorMaps,
  System.Actions;

type
  TfrmProbCalc = class(TForm)
    FrameDist1: TFrameDist;             
    PageControl1: TPageControl;                   
    TabSheetResults: TTabSheet;                               
    TabSheetChart: TTabSheet;
    MemoResults: TMemo;
    Chart1: TChart;
    Series1: TLineSeries;
    Panel2: TPanel;
    Panel3: TPanel;
    CheckBox3: TCheckBox;
    CheckBox4: TCheckBox;
    ActionManager1: TActionManager;
    aICDF: TAction;
    aCalc: TAction;
    aClearAll: TAction;
    ActionToolBar1: TActionToolBar;
    ActionToolBar2: TActionToolBar;
    aPDF: TAction;
    aCDF: TAction;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure aAdditionalInfoExecute(Sender: TObject);
    procedure Chart1AfterDraw(Sender: TObject);
    procedure FrameDist1RadioGroupOutClick(Sender: TObject);
    procedure aCalcExecute(Sender: TObject);
    procedure aICDFExecute(Sender: TObject);
    procedure aClearAllExecute(Sender: TObject);
    procedure aClearAllUpdate(Sender: TObject);
    procedure aPDFExecute(Sender: TObject);
    procedure aCDFExecute(Sender: TObject);
    procedure CheckBox4Click(Sender: TObject);
  private
    { Private declarations }
    Dist: TProbDistribution;
  public
    { Public declarations }
  end;

var
  frmProbCalc: TfrmProbCalc;

implementation

Uses ClipBrd;

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

procedure TfrmProbCalc.FormCreate(Sender: TObject);
begin
  { Setup distribution }
  Dist := TProbDistribution.Create;
  Dist.LB := 0;
  Dist.UB := 1;
  Dist.NumPoints := 50;
  { temporary, will be replaced with true editor later }
  FrameDist1.Distribution := Dist;
  FrameDist1.ListBoxDist.ItemIndex := Ord(Dist.DistType);
end;

procedure TfrmProbCalc.FormDestroy(Sender: TObject);
begin
  Dist.Free;
end;

procedure TfrmProbCalc.aAdditionalInfoExecute(Sender: TObject);
begin
  Chart1.Repaint;
end;

procedure TfrmProbCalc.Chart1AfterDraw(Sender: TObject);
var i,xstart,ystart,linc: Integer;
    tstr: String;
begin
  if CheckBox3.Checked then
  with Chart1, Chart1.Canvas do
  begin
    Font.Size := Title.Font.Size;
    linc := Round(Font.Size*1.3);
    xstart := ChartRect.Left + 30;
    ystart := ChartRect.Top + 30;
    tstr :=Dist.Caption+' distribution';
    TextOut(xstart,ystart,tstr);
    inc(ystart,linc*2);
    TextOut(xstart,ystart,'parameters:');
    Inc(ystart,linc);
    for i := 0 to Dist.DistParams.Count -1 do
    begin
      tstr := Dist.DistParams[i].Name+':'+FloatToStrF(Dist.DistParams[i].Value,ffFixed,6,6);
      TextOut(xstart,ystart,tstr);
      Inc(ystart,linc);
    end;
    Inc(ystart,linc);
    TextOut(xstart,ystart,'Plot range:'+FloatToStrF(Dist.LB,ffFixed,6,6)+' to '+FloatToStrF(Dist.UB,ffFixed,6,6));
  end;
end;

procedure TfrmProbCalc.CheckBox4Click(Sender: TObject);
begin
  Chart1.Repaint;
end;

procedure TfrmProbCalc.FrameDist1RadioGroupOutClick(Sender: TObject);
begin
  FrameDist1.RadioGroupOutClick(Sender);
  TabSheetChart.TabVisible := (aPDF.Checked) and (FrameDist1.RadioGroupOut.ItemIndex = 0);
end;

procedure TfrmProbCalc.aCalcExecute(Sender: TObject);

    function ResToStr(res: TSample): String;
    begin
      Result := SampleToStr(res,6,6);
    end;
    function Val2Str(x,val:TSample): String;
    begin
      Result := ResToStr(x)+#9+#9+ResToStr(val);
    end;
    procedure AddPointToSeries(aSeries:TChartSeries; x,y: TSample);
    begin
      if Not(IsNAN(x) or IsNan(y) or IsInf(x) or IsInf(y)) then
        aSeries.AddXY(x,y);
    end;

var i: Integer;
    calcstep,parvalue,yval: TSample;
    clb,cub: TSample;
begin
  if CheckBox4.Checked then aClearAllExecute(Self);
  Chart1.Title.Text.Clear;

  With MemoResults.Lines do
  begin
    Add(Dist.Caption+ ' distribution');
    for i := 0 to Dist.DistParams.Count -1 do
      Add(Dist.DistParams[i].Name + '= ' + SampleToStr(Dist.DistParams[i].Value,6,6));

    if aPDF.Checked then
    begin
      Chart1.Title.Text.Add('Probability density function');
      Chart1.Axes.Left.Title.Caption := 'PDF';
      Chart1.Axes.Bottom.Title.Caption := 'x';
      Add('Calculating probability density function (PDF)');
      Add('x'+#9+#9+'PDF(x)')
    end else if ACdf.Checked then
    begin
      Chart1.Title.Text.Add('Cumulative distribution function');
      Chart1.Axes.Left.Title.Caption := 'CDF';
      Chart1.Axes.Bottom.Title.Caption := 'x';
      Add('Calculating cumulative distribution function (CDF)');
      Add('x'+#9+#9+'CDF(x)')
    end else
    begin
      Chart1.Axes.Left.Title.Caption := 'PPF';
      Chart1.Axes.Left.Title.Caption := 'PPF';
      Chart1.Axes.Bottom.Title.Caption := 'p';
      Add('Calculating point percent function (PPF)');
      Add('p'+#9+#9+'x(p)')
    end;
    Add('---------------------------------------------');
    if FrameDist1.RadioGroupOut.ItemIndex = 0 then { single point }
    begin
      if (aICDF.Checked) then Add(Val2Str(Dist.p.Value,Dist.PPF))
      else if aPDF.Checked then Add(Val2Str(Dist.X.Value,Dist.PDF))
      else if aCDF.Checked then Add(Val2Str(Dist.X.Value,Dist.CDF))
    end
    else { range }
    begin
      clb := Dist.LB;
      cub := Dist.UB;
      calcstep := (cub-clb)/Dist.NumPoints;
      for i := 0 to Dist.NumPoints -1 do
      begin
        if aICDF.Checked then parvalue := 0+i/Dist.NumPoints
        else parvalue := clb + i*calcstep;
        yval := 0.0;
        if aPDf.Checked then yval := Dist.CalculatePDF(parvalue)
        else if ACDf.Checked then yval := Dist.CalculateCDF(parvalue)
        else if aICDF.Checked then yval := Dist.CalculatePPF(parvalue);

        Add(Val2Str(parvalue,yval));
        AddPointToSeries(Chart1.Series[0],parvalue,yval);
      end
    end;
    Add('---------------------------------------------');
  end;
end;

procedure TfrmProbCalc.aCDFExecute(Sender: TObject);
begin
  TabSheetChart.TabVisible := (FrameDist1.RadioGroupOut.ItemIndex <> 0);
  aCalcExecute(Self);
end;

procedure TfrmProbCalc.aPDFExecute(Sender: TObject);
begin
  TabSheetChart.TabVisible := (FrameDist1.RadioGroupOut.ItemIndex <> 0);
  aCalcExecute(Self);
end;

procedure TfrmProbCalc.aICDFExecute(Sender: TObject);
var tstr: String;
begin
TabSheetChart.TabVisible := (FrameDist1.RadioGroupOut.ItemIndex <> 0);
  if FrameDist1.RadioGroupOut.ItemIndex = 0 then
  begin
    tStr := InputBox('Set p','Valid range: 0<=p<=1',FloatToStrF(Dist.p.Value,ffFixed,6,6));
    try
      Dist.p.Value := StrToFloat(tstr);
    except
      raise
    end;
  end;
  aCalcExecute(Self);
end;

procedure TfrmProbCalc.aClearAllExecute(Sender: TObject);
var i: Integer;
begin
  MemoResults.Lines.Clear;
  for i := 0 to Chart1.SeriesCount -1 do Chart1.Series[i].Clear;
end;

procedure TfrmProbCalc.aClearAllUpdate(Sender: TObject);
begin
  aClearAll.Enabled := MemoResults.Lines.Count > 0;
end;

initialization
  RegisterClass(TfrmProbCalc);

end.

File: MtxVec.CPP/DewProbCalc.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#include <math.h>
#pragma hdrstop

#include "DewProbCalc.h"
#include <Vcl.Clipbrd.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "ProbEditor"
#pragma resource "*.dfm"
TfrmProbCalc *frmProbCalc;
//---------------------------------------------------------------------------
__fastcall TfrmProbCalc::TfrmProbCalc(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::FormCreate(TObject *Sender)
{
  Dist = new TProbDistribution();
  Dist->LB = 0;
  Dist->UB = 1;
  Dist->NumPoints = 50;
  FrameDist1->Distribution = Dist;
  FrameDist1->ListBoxDist->ItemIndex = (int)Dist->DistType;
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::FormDestroy(TObject *Sender)
{
  delete Dist;
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::aAdditionalInfoExecute(TObject *Sender)
{
  Chart1->Repaint();
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::Chart1AfterDraw(TObject *Sender)
{
  if (CheckBox3->Checked)
  {
    Chart1->Canvas->Font->Size = Chart1->Title->Font->Size;
    int linc = Math387::RoundToInt(Chart1->Canvas->Font->Size * 1.3);
    int xstart = Chart1->ChartRect.Left + 30;
    int ystart = Chart1->ChartRect.Top + 30;
    AnsiString tstr = Dist->Caption + " distribution";
    Chart1->Canvas->TextOut(xstart, ystart, tstr);
    ystart += linc * 2;
    Chart1->Canvas->TextOut(xstart, ystart, "parameters:");
    ystart += linc;
    for (int i = 0; i < Dist->DistParams->Count; i++)
    {
      tstr = Dist->DistParams->Items[i]->Name + ":" +
             FloatToStrF(Dist->DistParams->Items[i]->Value, ffFixed, 6, 6);
      Chart1->Canvas->TextOut(xstart, ystart, tstr);
      ystart += linc;
    }
    ystart += linc;
    Chart1->Canvas->TextOut(xstart, ystart,
        "Plot range:" + FloatToStrF(Dist->LB, ffFixed, 6, 6) +
        " to " + FloatToStrF(Dist->UB, ffFixed, 6, 6));
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::CheckBox4Click(TObject *Sender)
{
  Chart1->Repaint();
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::FrameDist1RadioGroupOutClick(TObject *Sender)
{
  FrameDist1->RadioGroupOutClick(Sender);
  TabSheetChart->TabVisible = (aPDF->Checked) && (FrameDist1->RadioGroupOut->ItemIndex == 0);
}
//---------------------------------------------------------------------------
static AnsiString ResToStr(double res) { return SampleToStr(res, 6, 6); }
static AnsiString Val2Str(double x, double val) { return ResToStr(x) + "\t\t" + ResToStr(val); }
static void AddPointToSeries(TChartSeries *aSeries, double x, double y)
{
  if (!(IsNan(x) || IsNan(y) || IsInf(x) || IsInf(y)))
    aSeries->AddXY(x, y);
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::aCalcExecute(TObject *Sender)
{
  if (CheckBox4->Checked) aClearAllExecute(this);
  Chart1->Title->Text->Clear();

  MemoResults->Lines->Add(Dist->Caption + " distribution");
  for (int i = 0; i < Dist->DistParams->Count; i++)
    MemoResults->Lines->Add(Dist->DistParams->Items[i]->Name + "= " +
                            SampleToStr(Dist->DistParams->Items[i]->Value, 6, 6));

  if (aPDF->Checked)
  {
    Chart1->Title->Text->Add("Probability density function");
    Chart1->Axes->Left->Title->Caption = "PDF";
    Chart1->Axes->Bottom->Title->Caption = "x";
    MemoResults->Lines->Add("Calculating probability density function (PDF)");
    MemoResults->Lines->Add("x\t\tPDF(x)");
  }
  else if (aCDF->Checked)
  {
    Chart1->Title->Text->Add("Cumulative distribution function");
    Chart1->Axes->Left->Title->Caption = "CDF";
    Chart1->Axes->Bottom->Title->Caption = "x";
    MemoResults->Lines->Add("Calculating cumulative distribution function (CDF)");
    MemoResults->Lines->Add("x\t\tCDF(x)");
  }
  else
  {
    Chart1->Axes->Left->Title->Caption = "PPF";
    Chart1->Axes->Bottom->Title->Caption = "p";
    MemoResults->Lines->Add("Calculating point percent function (PPF)");
    MemoResults->Lines->Add("p\t\tx(p)");
  }
  MemoResults->Lines->Add("---------------------------------------------");

  if (FrameDist1->RadioGroupOut->ItemIndex == 0)
  {
    // single point
    if (aICDF->Checked) MemoResults->Lines->Add(Val2Str(Dist->p->Value, Dist->PPF));
    else if (aPDF->Checked) MemoResults->Lines->Add(Val2Str(Dist->X->Value, Dist->PDF));
    else if (aCDF->Checked) MemoResults->Lines->Add(Val2Str(Dist->X->Value, Dist->CDF));
  }
  else
  {
    // range
    double clb = Dist->LB;
    double cub = Dist->UB;
    double calcstep = (cub - clb) / Dist->NumPoints;
    for (int i = 0; i < Dist->NumPoints; i++)
    {
      double parvalue;
      if (aICDF->Checked) parvalue = 0 + (double)i / Dist->NumPoints;
      else parvalue = clb + i * calcstep;
      double yval = 0.0;
      if (aPDF->Checked) yval = Dist->CalculatePDF(parvalue);
      else if (aCDF->Checked) yval = Dist->CalculateCDF(parvalue);
      else if (aICDF->Checked) yval = Dist->CalculatePPF(parvalue);

      MemoResults->Lines->Add(Val2Str(parvalue, yval));
      AddPointToSeries(Chart1->Series[0], parvalue, yval);
    }
  }
  MemoResults->Lines->Add("---------------------------------------------");
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::aCDFExecute(TObject *Sender)
{
  TabSheetChart->TabVisible = (FrameDist1->RadioGroupOut->ItemIndex != 0);
  aCalcExecute(this);
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::aPDFExecute(TObject *Sender)
{
  TabSheetChart->TabVisible = (FrameDist1->RadioGroupOut->ItemIndex != 0);
  aCalcExecute(this);
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::aICDFExecute(TObject *Sender)
{
  TabSheetChart->TabVisible = (FrameDist1->RadioGroupOut->ItemIndex != 0);
  if (FrameDist1->RadioGroupOut->ItemIndex == 0)
  {
    AnsiString tStr = InputBox("Set p", "Valid range: 0<=p<=1",
                               FloatToStrF(Dist->p->Value, ffFixed, 6, 6));
    Dist->p->Value = StrToFloat(tStr);
  }
  aCalcExecute(this);
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::aClearAllExecute(TObject *Sender)
{
  MemoResults->Lines->Clear();
  for (int i = 0; i < Chart1->SeriesCount(); i++)
    Chart1->Series[i]->Clear();
}
//---------------------------------------------------------------------------
void __fastcall TfrmProbCalc::aClearAllUpdate(TObject *Sender)
{
  aClearAll->Enabled = MemoResults->Lines->Count > 0;
}
//---------------------------------------------------------------------------
